sdk/program/src/poseidon.rs

```diff
@@ -37,25 +37,6 @@ pub enum PoseidonSyscallError {
     Unexpected,
 }
 
-impl From<u64> for PoseidonSyscallError {
```

Comment:
I don't think this is enough? Look at the `hashv` function starting from line 183 - in both cases (`#[cfg(os = "solana")]` and `#[cfg(not(os = "solana"))]`) the errors remained untouched and you probably want to apply the same change there as you did in the alt_bn part.

Reply:
I don't understand. For the curve ops, at the lib level we still return a typed error. The syscall then masks that and returns Ok(1). Same for Poseidon?

Reply:
The tests also seem related? I don't think we should hide what type of error is happening to developers, we should just simplify the syscall return code (because it affects consensus). There's currently 0 tests on syscalls error cases (not just alt_bn128, all syscalls), we're addressing this issue in independent work.

Reply:
>I don't think we should hide what type of error is happening to developers, we should just simplify the syscall return code (because it affects consensus).

I don't think it's possible if we are unifying the error on the syscall side? We either return distinct error codes which can be converted into distinct error variants, or we return one error code and then just have one error variant being always returned. Or am I missing anything? If we return always `Err(1)`, probably the only way to notify developers of errors are explicit logs.

Currently, and even after your changes, the library code calls the syscalls and casts its error into `PoseidonSyscallError`:

https://github.com/anza-xyz/agave/blob/5447a73b5ba006e3e26ab23f11374ce5ec7840a3/sdk/program/src/poseidon.rs#L251

From my understanding you run this against a validator with always returns `Err(1)`, it will get always converted into `PoseidonSyscallError::InvalidParameters`, if I'm not missing anything?

Wouldn't it be more correct to return `PoseidonSyscallError::Unexpected` and then, in future, just remove all the other variants, as they wouldn't be used?

Reply:
I understand the problem, but I don't know what the correct solution would be. See, e.g., for an identical case:
https://github.com/anza-xyz/agave/blob/master/programs/bpf_loader/src/syscalls/mod.rs#L730
https://github.com/anza-xyz/agave/blob/8887cd19a1a86f61cfa44fb41a6e528e8857bc59/sdk/program/src/pubkey.rs#L629

Reply:
Yeah, I think the best would be doing something like this in the library code:

```rust
    // Call via a system call to perform the calculation.
    #[cfg(target_os = "solana")]
    {
        let mut hash_result = [0; HASH_BYTES];
        let result = unsafe {
            crate::syscalls::sol_poseidon(
                parameters.into(),
                endianness.into(),
                vals as *const _ as *const u8,
                vals.len() as u64,
                &mut hash_result as *mut _ as *mut u8,
            )
        };

        match result {
            0 => Ok(PoseidonHash::new(hash_result)),
            e => Err(PoseidonSyscallError::Unexpected), // <- HERE
        }
    }
```

Then after some time, when the feature gate is well adopted, we could just remove all other error variants from `PoseidonSyscallError`. What do you think?

Reply:
I just had a convo with the team. Of course, everyone has a different opinion :)

I think we all agree on the following:
1. The pub key code that I referred is old and shouldn't be used as an example
2. Having an inconsistent API is bad

If we leave the PR as is (and/or with your proposed change), we'll have to change it again soon, which seems suboptimal.

So the proposal is to have a single `XxxSyscallError` (e.g., struct instead of enum). And do all the necessary changes. For Poseidon specifically we should prob return `Err` like in all other hash functions, as the error(s) are not really errors that the program may recover from.

@vadorovsky if you agree I'll go ahead and make the changes. If not we should prob discuss this further e.g. in #vitual-machine.